src/Optical/Lenses.jl

```diff
@@ -2,123 +2,376 @@
-    @assert semidiameter > zero(S)
+function AsphericLens(
+    insidematerial::AbstractGlass,
+    frontvertex::T,
+    frontradius::T,
+    frontconic::T,
+    frontaspherics::Union{Nothing,Vector{Tuple{Int,T}}},
+    backradius::T,
+    backconic::T,
+    backaspherics::Union{Nothing,Vector{Tuple{Int,T}}},
+    thickness::T,
+    semidiameter::T;
+    lastmaterial::AbstractGlass = Air,
+    nextmaterial::AbstractGlass = Air,
+    frontsurfacereflectance::T = zero(T),
+    backsurfacereflectance::T = zero(T),
+    nsamples::Int = 17,
+    frontdecenter::Tuple{T,T} = (zero(T), zero(T)),
+    backdecenter::Tuple{T,T} = (zero(T), zero(T)),
+    interfacemode::InterfaceMode = ReflectOrTransmit
+) where {T<:Real}
+    @assert semidiameter > zero(T)
     @assert !isnan(frontradius)
 
-    frontdecenter_l = abs(frontdecenter[1]) > eps(S) && abs(frontdecenter[2]) > eps(S) ? sqrt(frontdecenter[1]^2 + frontdecenter[2]^2) : zero(S)
-    backdecenter_l = abs(backdecenter[1]) > eps(S) && abs(backdecenter[2]) > eps(S) ? sqrt(backdecenter[1]^2 + backdecenter[2]^2) : zero(S)
-
-    if isinf(frontradius) && (frontaspherics === nothing)
-        #planar
-        lens_front = leaf(Plane(SVector{3,S}(0, 0, 1), SVector{3,S}(0, 0, frontvertex), vishalfsizeu = semidiameter, vishalfsizev = semidiameter, interface = interface = opticinterface(S, insidematerial, lastmaterial, frontsurfacereflectance, interfacemode)))
-    elseif frontconic == zero(S) && (frontaspherics === nothing)
-        #spherical
-        if frontradius < zero(S)
-            ϕmax = NaNsafeasin(min(abs((min(semidiameter, abs(frontradius)) + frontdecenter_l) / frontradius), one(S))) + S(π / 50)
-            lens_front = leaf(SphericalCap(abs(frontradius), ϕmax, SVector{3,S}(0, 0, 1), SVector{3,S}(frontdecenter[1], frontdecenter[2], frontvertex), interface = opticinterface(S, insidematerial, lastmaterial, frontsurfacereflectance, interfacemode)))
+    frontdecenter_l = (
+        abs(frontdecenter[1]) > eps(T) && abs(frontdecenter[2]) > eps(T) ?
+        sqrt(frontdecenter[1]^2 + frontdecenter[2]^2) :
+        zero(T)
+    )
+    backdecenter_l = (
+        abs(backdecenter[1]) > eps(T) && abs(backdecenter[2]) > eps(T) ?
+        sqrt(backdecenter[1]^2 + backdecenter[2]^2) :
+        zero(T)
+    )
+
+    # front lens
+    frontinterface = opticinterface(T, insidematerial, lastmaterial, frontsurfacereflectance, interfacemode)
```

Comment:
Other than interface, is there any reason to treat the front and back lenses separately? I'm reading and comparing both halves of the code to spot the differences now.

Reply:
Front and back could have different reflectance - in the future they may have totally different interfaces due to different coatings being present on each side (very common)

Reply:
Vague wording from me here - I was trying to get an understanding of how much of the somewhat duplicated code is unique between the front/back lenses. My motivation for this is that I'll probably be splitting the lens code into

1. Various functions for creating a (front or back) lens surface of each surface type
2. A function for combining each of these surfaces into a CSG lens

I wanted to double-check that the only significant differences between front and back would be their interfaces and a z-translation. Is this correct?